Replaced the method of execution in the TaskDispatcher.

ChubbyQuokka opened the pull request:
**~~WIP, I need to test that this is Thread-Safe before fully merging.~~**
**This is now ready to merge.**

This is just a small change to help prevent a long list of actions from keeping a lock on the object too long (possibly stalling another `Thread` attempting to obtain the lock) by using a set of `Queue` buffers. I can also change the back to using the `List` instead of a `Queue` if you guys would like.

This also fixes a bug when the `TaskDispatcher` is shutdown where the `QueuedMainFixedActions` is not cleared and the `QueuedAyncActions` is cleared twice.

fr34kyn01535 commented on Rocket.API/Utils/TaskDispatcher.cs:
```diff
@@ -8,9 +8,16 @@ namespace Rocket.API.Utils
-        private static readonly List<Action> QueuedAsyncActions = new List<Action>();
+
+        private static readonly Queue<Action> QueuedMainActions = new Queue<Action>();
+        private static readonly Queue<Action> QueuedMainActionsBuffer = new Queue<Action>();
```
Why is there a second queue for queued main actions?

ChubbyQuokka replied:
The second set of queues act as a buffer to allow the thread requesting the action on the other thread to allow execution even if the other thread is working on its current queue.

ChubbyQuokka replied:
This would mainly help prevent a thread from stalling when it tries to add something to the queue of a different thread that's performing a heavily I/O based function that could take longer than a few milliseconds.

fr34kyn01535 replied:
Well then don't lock it that long I suppose. Dequeue what's in the queue and leave the lock, then do the action. There is no need for a second collection.

ChubbyQuokka replied:
Without the second collection, it would have to re-enter the lock multiple times per tick or only execute one action per tick.

fr34kyn01535 replied:
Why not keep it as list, enter the lock, copy the list to a local variable in the main thread, leave the lock and handle the local list.

ChubbyQuokka replied:
By using a local list, wouldn't the object have to be re instantiated every tick? At least on the main thread where the method does eventually exit to allow the tread to handle other things. I can change them back to lists also, but a queue seems like it is more designed for a job like this.

fr34kyn01535 replied:
If you intend to fetch all entries anyway i suppose a queue is not making the job easier. One could also implement a custom collection that returns new lists when the thread is asking for the data & clears its contents automatically, not blocking the list in the time the resulting collection is iterated.

ChubbyQuokka commented:
@fr34kyn01535 Alright, I switched to using lists.

I also added a `volatile bool` to allow the main `Thread` to tell the secondary `Thread` to stop iterating.

I have also modified the `Shutdown()` method to prevent it from restarting the `Thread` and also to join the `Thread` to prevent any corruption to any I/O based operations it's currently performing.

ChubbyQuokka commented:
I guess with the IOC branch, I'll go ahead and close this.
